src/TypeSystem/Inferrer/TypeInferrer.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/**
+ * PackageFactory.ComponentEngine - Universal View Components for PHP
+ *   Copyright (C) 2022 Contributors of PackageFactory.ComponentEngine
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+declare(strict_types=1);
+
+namespace PackageFactory\ComponentEngine\TypeSystem\Inferrer;
+
+use PackageFactory\ComponentEngine\Definition\BinaryOperator;
+use PackageFactory\ComponentEngine\Parser\Ast\BinaryOperationNode;
+use PackageFactory\ComponentEngine\Parser\Ast\ExpressionNode;
+use PackageFactory\ComponentEngine\Parser\Ast\IdentifierNode;
+use PackageFactory\ComponentEngine\Parser\Ast\NullLiteralNode;
+use PackageFactory\ComponentEngine\TypeSystem\ScopeInterface;
+
+/**
+ * This class handles the analysis of identifier types that are used in a condition
+ * and based on the requested branch: truthy or falsy, will predict the types a variable will have in the respective branch
+ * so it matches the expected runtime behaviour
+ *
+ * For example given this expression: `nullableString ? "nullableString is not null" : "nullableString is null"` based on the condition `nullableString`
+ * It will infer that in the truthy context nullableString is a string while in the falsy context it will infer that it is a null
+ *
+ * The structure is partially inspired by phpstan
+ * https://github.com/phpstan/phpstan-src/blob/07bb4aa2d5e39dafa78f56c5df132c763c2d1b67/src/Analyser/TypeSpecifier.php#L111
+ */
+class TypeInferrer
+{
+    public function __construct(
+        private readonly ScopeInterface $scope
+    ) {
+    }
+
+    public function inferTypesInCondition(ExpressionNode $conditionNode, TypeInferrerContext $context): InferredTypes
```

Comment:
```suggestion
    public function narrowTypesOfSymbolsIn(ExpressionNode $expressionNode): TypeMap
```

Imho, we do not need to constraint this particular algorithm to ternary condition nodes. Type narrowing can be done given any expression.

Furthermore, as far as I understand, `TypeInferrerContext` is supposed to inform this method of whether we assume the ternary condition to be truthy or falsy. I don't believe we need to make this distinction. It is sufficient to perform the narrowing algorithm under the assumption that `$expressionNode` evaluates to be truthy. Given the result of that, for the assumption that `$expressionNode` evaluates to be falsy the opposite must be true (which means that `UnionType` may need another method to build the difference with a given `UnionType`).

For performance it may be wise to have this class cache the result of this method in memory and let both `TernaryBranchScope`s share one instance.

Reply:
> Furthermore, as far as I understand, TypeInferrerContext is supposed to inform this method of whether we assume the ternary condition to be truthy or falsy.

Exactly ;)

>  I don't believe we need to make this distinction. It is sufficient to perform the narrowing algorithm under the assumption that $expressionNode evaluates to be truthy. Given the result of that, for the assumption that $expressionNode evaluates to be falsy the opposite must be true (which means that UnionType may need another method to build the difference with a given UnionType).

Hmm i was convinced by this approach at first but i have the feeling that important informations will be lost when narrowing the true branch, and its not (with out current TypeMap) not possible to invert it correctly:

`foo ? true : false`

if `foo` is of type `string|null` the true ast will know its `string` and inverting is not possible since we dont know yet if this was ever a nullable type - we cant just make it null for the false branch, as it might not have been nullable in the first place. Also the idea was copied from phpstan and in my head it makes perfectly sense when the expression contains complex logic, but i cant bring up another example right now ^^

Either way id like to propose to stick to it for know and refactor later if we dont need it or find it "too slow".

Reply:
You are very right! I had a thinking error there :sweat_smile: 

My statement
> Given the result of [the narrowing algorithm under the assumption that $expressionNode evaluates to be truthy], for the assumption that $expressionNode evaluates to be falsy the opposite must be true

can easily be falsified:

```typescript
foo: null|string
bar: null|string

foo && bar
  ? (/* foo and bar MUST be of type string */)
  : (/* foo and bar MAY each be of type null|string */)
```

So, it is inevitable to make the distinction. However, I'd still like to suggest two changes here:

1. Rename `TypeInferrerContext` to `Truthiness`
2. Move the parameter that is currently called `$context` to the constructor (and probably rename it to something like `$assumedTruthiness` - not sure right now whether that's precise enough), so the resulting method `narrowTypesOfSymbolsIn` becomes unary

Wdyt?

Reply:
Adjusted ;)

Discussed using instance property for `$assumedTruthiness` with @fcool and he found it also a good idea, so we did it ^^